server/reload_test.go

```diff
@@ -1787,6 +1788,519 @@ func TestConfigReloadClusterName(t *testing.T) {
 	}
 }
 
+func TestConfigClusterMembershipReload(t *testing.T) {
+	//
+	// [A] starts in dynamic mode.
+	//
+	s1, _, conf1 := runReloadServerWithContent(t, []byte(`
+	listen: "0.0.0.0:-1"
+        server_name: "A"
+
+	cluster: {
+          listen: "0.0.0.0:-1"
+	}
+	`))
+	defer os.Remove(conf1)
+	defer s1.Shutdown()
+
+	got := s1.ClusterName()
+	if got == "" {
+		t.Fatalf("Expected update clustername to be set dynamically")
+	}
+
+	//
+	// [A] joins AB cluster with explicit name.
+	//
+	reloadUpdateConfig(t, s1, conf1, `
+        listen: "0.0.0.0:-1"
+        server_name: "A"
+
+        cluster: {
+          name: "AB"
+          listen: "0.0.0.0:-1"
+        }
+        `)
+	if s1.ClusterName() != "AB" {
+		t.Fatalf("Expected update clustername of \"AB\", got %q", s1.ClusterName())
+	}
+
+	//
+	// [B] joins AB cluster with explicit name.
+	//
+	template := fmt.Sprintf(`
+	listen: "0.0.0.0:-1"
+        server_name: "B"
+
+	cluster: {
+          name: "AB"
+          listen: "0.0.0.0:-1"
+          routes: [ nats://localhost:%d ] # Route to A
+	}
+	`, s1.ClusterAddr().Port)
+
+	s2, _, conf2 := runReloadServerWithContent(t, []byte(template))
+	defer os.Remove(conf2)
+	defer s2.Shutdown()
+	checkClusterFormed(t, s1, s2)
+
+	//
+	// [A] client sends request to [B] client and should be able to respond.
+	//
+	nc1, err := nats.Connect(fmt.Sprintf("nats://%s:%d", "127.0.0.1", s1.Addr().(*net.TCPAddr).Port))
+	if err != nil {
+		t.Fatalf("Error creating client: %v", err)
+	}
+	defer nc1.Close()
+
+	nc2, err := nats.Connect(fmt.Sprintf("nats://%s:%d", "127.0.0.1", s2.Addr().(*net.TCPAddr).Port))
+	if err != nil {
+		t.Fatalf("Error creating client: %v", err)
+	}
+	defer nc2.Close()
+	nc1.Subscribe("test", func(m *nats.Msg) {
+		m.Respond([]byte("pong"))
+	})
+	nc1.Flush()
+	_, err = nc2.Request("test", []byte("ping"), 2*time.Second)
```

Comment:
Technically, nc1.Flush() does not guarantee that the subscription interest has made it to server B, so for the request to never timeout you would need to check subject interest on B: check the helper: `func checkSubInterest(t *testing.T, s *Server, accName, subject string, timeout time.Duration)`.
Here it would be something like:
```
checkSubInterest(t, s2, globalAccountName, "test", time.Second)
```